File: scripts/darwin/itunes.js
const app = Application.currentApplication();
app.includeStandardAdditions = true;
const version = Number(app.doShellScript("sw_vers -productVersion| cut -d '.' -f 1-2"));
const itunes = version < 10.15 ? Application("iTunes") : Application("Music");

const map = (array, callback) => {
  const objArray = [];
  for (let i = 0; i < array.length; i++) {
    objArray.push(callback(array[i], i));
  }
  return objArray;
};

const getValue = (value) => {
  return (property) => {
    try {
      return value[property]();
    } catch (e) {
      return null;
    }
  };
};

/**
 * get Item properties
 * @param {*} item
 * @returns
 */
const getItem = (item) => {
  const getItemValue = getValue(item);
  return {
    class: getItemValue("class"),
    container: getItemValue("container"),
    id: getItemValue("id"),
    index: getItemValue("index"),
    name: getItemValue("name"),
    persistentId: getItemValue("persistentID"),
    properties: getItemValue("properties"),
  };
};

const getAirPlayDevices = (device) => {
  const getDeviceValue = getValue(device);
  return Object.assign(getItem(device), {
    active: getDeviceValue("active"),
    available: getDeviceValue("available"),
    kind: getDeviceValue("kind"),
    networkAddress: getDeviceValue("networkAddress"),
    protected: getDeviceValue("protected"),
    selected: getDeviceValue("selected"),
    supportsAudio: getDeviceValue("supportsAudio"),
    supportsVideo: getDeviceValue("supportsVideo"),
    soundVolume: getDeviceValue("soundVolume"),
  });
};

const getEncoder = (encoder) => {
  const getEncoderValue = getValue(encoder);
  return Object.assign(getItem(encoder), {
    format: getEncoderValue("format"),
  });
};

const getEQPreset = (eqPreset) => {
  const getEQPresetValue = getValue(eqPreset);
  return Object.assign(getItem(eqPreset), {
    band1: getEQPresetValue("band1"),
    band2: getEQPresetValue("band2()"),
    band3: getEQPresetValue("band3"),
    band4: getEQPresetValue("band4"),
    band5: getEQPresetValue("band5"),
    band6: getEQPresetValue("band6"),
    band7: getEQPresetValue("band7"),
    band8: getEQPresetValue("band8"),
    band9: getEQPresetValue("band9"),
    band10: getEQPresetValue("band10"),
    modifiable: getEQPresetValue("modifiable"),
    preamp: getEQPresetValue("preamp"),
    updateTracks: getEQPresetValue("updateTracks"),
  });
};

const getPlaylist = (playlist) => {
  const getPlaylistValue = getValue(playlist);

  const tracks = toArray(playlist.tracks).map((track) => getTrack(track));
  const artworks = toArray(playlist.artworks).map((artwork) =>
    getArtwork(artwork)
  );

  return Object.assign(getItem(playlist), {
    tracks,
    artworks,
    description: getPlaylistValue("description"),
    disliked: getPlaylistValue("disliked"),
    duration: getPlaylistValue("duration"),
    name: getPlaylistValue("name"),
    loved: getPlaylistValue("loved"),
    parent: getPlaylistValue("parent"),
    shuffle: getPlaylistValue("shuffle"),
    size: getPlaylistValue("size"),
    songRepeat: getPlaylistValue("songRepeat"),
    specialKind: getPlaylistValue("specialKind"),
    time: getPlaylistValue("time"),
    visible: getPlaylistValue("visible"),
  });
};

const getArtwork = (artwork) => {
  const getArtworkValue = getValue(artwork);
  return Object.assign(getItem(artwork), {
    data: getArtworkValue("data"),
    description: getArtworkValue("description"),
    downloaded: getArtworkValue("downloaded"),
    format: getArtworkValue("format"),
    kind: getArtworkValue("kind"),
    rawData: getArtworkValue("rawData"),
  });
};

const getTrack = (track) => {
  const getTrackValue = getValue(track);
  const artworks =
    getTrackValue("artworks") == null
      ? []
      : toArray(track.artworks).map((artwork) => {
          return getArtwork(artwork);
        });

  return Object.assign(getItem(track), getTrackValue("properties"), {
    artworks: artworks,
    lyrics: JSON.stringify(getTrackValue("lyrics"))
  });
};

const getVisual = (visual) => {
  return getItem(visual);
};

const getBrowserWindow = (browserWindow) => {
  const getBrowserWindowValue = getValue(browserWindow);
  return Object.assign(getItem(browserWindow), {
    selection: getBrowserWindowValue("selection"),
    view: getPlaylist(browserWindow),
  });
};

const getWindow = (window) => {
  const getWindowValue = getValue(window);
  return Object.assign(getItem(window), {
    bounds: getWindowValue("bounds"),
    closeable: getWindowValue("closeable"),
    collapseable: getWindowValue("collapseable"),
    collapseable: getWindowValue("collapseable"),
    fullScreen: getWindowValue("fullScreen"),
    position: getWindowValue("position"),
    resizable: getWindowValue("resizable"),
    visible: getWindowValue("visible"),
    zoomable: getWindowValue("zoomable"),
    zoomed: getWindowValue("zoomed"),
  });
};

const getEQWindow = (eqWindow) => {
  return getWindow(eqWindow);
};

const getMiniplayerWindow = (miniplayerWindow) => {
  return getWindow(miniplayerWindow);
};

const getPlaylistWindow = (playlistWindow) => {
  // TODO: selectionの配列を変換する
  return Object.assign(getWindow(playlistWindow), {
    selection: getTrack(playlistWindow.selection)["0"],
    view: getPlaylist(playlistWindow),
  });
};

const getSource = (source) => {
  const getSourceValue = getValue(source);

  // FIXME: 高速化
  const audioCDPlaylists = map(source.audioCDPlaylists, (audioCDPlaylist) =>
    getAudioCDPlaylist(audioCDPlaylist)
  );
  const libraryPlaylists = map(source.libraryPlaylists, libraryPlaylist => getLibraryPlayList(libraryPlaylist))
  const playlists = map(source.playlists, playlist => getPlaylist(playlist))
  const subscriptionPlaylists = map(
    source.subscriptionPlaylists,
    (subscriptionPlaylist) => getSubscriptionPlaylist(subscriptionPlaylist)
  );
  const userPlaylists = map(source.userPlaylists, userPlaylist => getUserPlaylist(userPlaylist))
  const radioTunerPlaylists = map(
    source.radioTunerPlaylists,
    (radioTunerPlaylist) => getRadioTunerPlaylist(radioTunerPlaylist)
  );

  return Object.assign(getItem(source), {
    audioCDPlaylists,
    libraryPlaylists,
    playlists,
    subscriptionPlaylists,
    userPlaylists,
    radioTunerPlaylists,
    capacity: getSourceValue("capacity"),
    freeSpace: getSourceValue("freeSpace"),
    kind: getSourceValue("kind"),
  });
};

const getVideoWindow = (videoWindow) => {
  return getWindow(videoWindow);
};

const getUserPlaylist = (userPlaylist) => {
  const getUserPlayListValue = getValue(userPlaylist);
  // FIXME: 高速化
  const fileTracks =
  getUserPlayListValue("fileTracks") == null
    ? []
    : map(userPlaylist.fileTracks, (fileTrack) => {
        return getFileTrack(fileTrack);
      });

  const urlTracks =
    getUserPlayListValue("urlTracks") == null
      ? []
      : map(userPlaylist.urlTracks, (urlTrack) => {
          return getUrlTrack(urlTrack);
        });

  const sharedTracks =
    getUserPlayListValue("sharedTracks") == null
      ? []
      : map(userPlaylist.sharedTracks, (sharedTrack) => {
          return getSharedTrack(sharedTrack);
        });

  return Object.assign(getPlaylist(userPlaylist), {
    fileTracks,
    urlTracks,
    sharedTracks,
    shared: getUserPlayListValue("shared"),
    smart: getUserPlayListValue("smart"),
    genius: getUserPlayListValue("genius"),
  });
};

const getFileTrack = (fileTrack) => {
  const getFileTrackValue = getValue(fileTrack);
  return Object.assign(getTrack(fileTrack), {
    location: getFileTrackValue("location"),
  });
};

const getURLTrack = (urlTrack) => {
  const getURLTrackValue = getValue(urlTrack);
  return Object.assign(getTrack(urlTrack), {
    address: getURLTrackValue("address"),
  });
};

const getSharedTrack = (sharedTrack) => {
  return getTrack(sharedTrack);
};

const getSubscriptionPlaylist = (subscriptionPlaylist) => {
  const getSubscriptionPlaylistValue = getValue(subscriptionPlaylist);
  // FIXME: 高速化
  const fileTracks =
  getSubscriptionPlaylistValue("fileTracks") == null
    ? []
    : map(subscriptionPlaylist.fileTracks, (fileTrack) => {
        return getFileTrack(fileTrack);
      });

  const urlTracks =
    getSubscriptionPlaylistValue("urlTracks") == null
      ? []
      : map(subscriptionPlaylist.urlTracks, (urlTrack) => {
          return getUrlTrack(urlTrack);
        });

  return {
    fileTracks,
    urlTracks,
  };
};

const getLibraryPlayList = (libraryPlayList) => {
  const fileTracks = map(libraryPlayList.fileTracks, (fileTrack) => getFileTrack(fileTrack));
  const urlTracks = map(libraryPlayList.urlTracks, (urlTrack) =>
    getUrlTrack(urlTrack)
  );
  const sharedTracks = map(libraryPlayList.sharedTracks, (sharedTrack) =>
    getSharedTrack(sharedTrack)
  );

  return Object.assign(getPlaylist(libraryPlayList), {
    fileTracks,
    urlTracks,
    sharedTracks,
  });
};

const getFolderPlaylist = (folderPlaylist) => {
  const getFolderPlaylistValue = getValue(folderPlaylist);

  const fileTracks = map(libraryPlayList.fileTracks, (fileTrack) => getFileTrack(fileTrack));
  const urlTracks = map(folderPlaylist.urlTracks, (urlTrack) =>
    getUrlTrack(urlTrack)
  );
  const sharedTracks = map(folderPlaylist.sharedTracks, (sharedTrack) =>
    getSharedTrack(sharedTrack)
  );

  return Object.assign(getUserPlaylist(folderPlaylist), {
    fileTracks,
    urlTracks,
    sharedTracks,
    shared: getFolderPlaylistValue("shared"),
    smart: getFolderPlaylistValue("smart"),
    genius: getFolderPlaylistValue("genius"),
  });
};

const getAudioCDTrack = (audioCDTrack) => {
  const getAudioCDTrackValue = getValue(audioCDTrack);
  return Object.assign(getTrack(audioCDTrack), {
    location: getAudioCDTrackValue("location"),
  });
};

const getAudioCDPlaylist = (audioCDPlaylist) => {
  const getAudioCDPlaylistValue = getValue(audioCDPlaylist);
  const audioCDTracks = map(audioCDPlaylist.audioCDTracks, (audioCDTrack) =>
    getAudioCDTrack(audioCDTrack)
  );

  return Object.assign(getPlaylist(audioCDPlaylist), {
    audioCDTracks,
    artist: getAudioCDPlaylistValue("artist"),
    compilation: getAudioCDPlaylistValue("compilation"),
    composer: getAudioCDPlaylistValue("composer"),
    discCount: getAudioCDPlaylistValue("discCount"),
    discNumber: getAudioCDPlaylistValue("discNumber"),
    genre: getAudioCDPlaylistValue("genre"),
    year: getAudioCDPlaylistValue("year"),
  });
};

const getRadioTunerPlaylist = (radioTunerPlaylist) => {
  const urlTracks = map(radioTunerPlaylist.urlTracks, (urlTrack) =>
    getUrlTrack(urlTrack)
  );
  return Object.assign(getPlaylist(radioTunerPlaylist), {
    urlTracks,
  });
};

/**
 * the currently selected AirPlay device(s)
 * @returns []airplayDevice
 */
const currentAirPlayDevices = (options) => {
  const devices = itunes.currentAirPlayDevices();
  return devices.map((device) => {
    return getAirPlayDevices(device);
  });
};

/**
 * the currently selected encoder (MP3, AIFF, WAV, etc.)
 * @returns Encoder
 */
const currentEncoder = (options) => {
  const encoder = itunes.currentEncoder();
  return getEncoder(encoder);
};

/**
 * the currently selected equalizer prese
 * @returns EQPreset
 */
const currentEQPreset = (options) => {
  const eqPreset = itunes.currentEQPreset();
  return getEQPreset(eqPreset);
};

/**
 * the playlist containing the currently targeted track
 * @returns playlist
 */
const currentPlaylist = (options) => {
  const playlist = itunes.currentPlaylist();
  return getPlaylist(playlist, options);
};

/**
 * the current targeted track
 * @returns track
 */
const currentTrack = (options) => {
  const track = itunes.currentTrack;
  return getTrack(track, options);
};

/**
 * the currently selected visual plug-in
 * @returns visual
 */
const currentVisual = (options) => {
  const visual = itunes.currentVisual();
  return getVisual(visual);
};

/**
 * main iTunes windows
 * @returns
 */
const browserWindow = (options) => {
  return map(itunes.browserWindows, (browserWindow) =>
    getBrowserWindow(browserWindow)
  );
};

const windows = () => {
  return map(itunes.windows, (window) => getWindow(window));
};

const encoders = (options) => {
  return map(itunes.encoders, (encoder) => getEncoder(encoder));
};

const eqPresets = (options) => {
  return map(itunes.eqPresets, (eqPreset) => getEQPreset(eqPreset));
};

const eqWindows = (options) => {
  return map(itunes.eqWindows, (eqWindow) => getEQWindow(eqWindow));
};

const miniplayerWindows = (options) => {
  return map(itunes.miniplayerWindows, (miniplayerWindow) =>
    getMiniplayerWindow(miniplayerWindow)
  );
};

const playlists = (options) => {
  return map(itunes.playlists, (playlist) => getPlaylist(playlist));
};

const playlistWindows = (options) => {
  return map(itunes.playlistWindows, (playlistWidow) =>
    getPlaylistWindow(playlistWidow)
  );
};

const sources = (options) => {
  return map(itunes.sources, (source) => getSource(source));
};

const tracks = (options) => {
  // FIXME: 処理速度が遅い
  return map(itunes.tracks, (track) => getTrack(track));
};

const videoWindows = (options) => {
  return map(itunes.videoWindows, (videoWindow) => getVideoWindow(videoWindow));
};

const visuals = (options) => {
  return map(itunes.visuals, (visual) => getVisual(visual));
};

const userPlaylists = (options) => {
  return map(itunes.userPlaylists, (userPlaylist) =>
    getUserPlaylist(userPlaylist)
  );
};

const subscriptionPlaylists = (options) => {
  return map(itunes.subscriptionPlaylists, (subscriptionPlaylist) =>
    getSubscriptionPlaylist(subscriptionPlaylist)
  );
};

const libraryPlaylists = (options) => {
  return map(itunes.libraryPlaylists, (libraryPlaylist) =>
    getLibraryPlayList(libraryPlaylist)
  );
};

const folderPlaylists = (options) => {
  return map(itunes.folderPlaylists, (folderPlaylist) =>
    getFolderPlaylists(folderPlaylist)
  );
};

const audioCDPlaylists = (options) => {
  return map(itunes.audioCDPlaylists, (audioCDPlaylist) =>
    getAudioCDPlaylist(audioCDPlaylist)
  );
};

const radioTunerPlaylists = (options) => {
  return map(itunes.radioTunerPlaylists, (radioTunerPlaylist) =>
    getRadioTunerPlaylist(radioTunerPlaylist)
  );
};

/**
 * return exit code
 * @param {*} code error code
 */
const returnError = (code = 1) => {
  ObjC.import("stdlib");
  $.exit(code);
};

function run(args) {
  const command = args[0];
  const options = {};
  args
    .slice(1)
    .filter((arg) => /^.+=.+$/.test(arg))
    .forEach((arg) => {
      const [key, value] = arg.split("=");
      options[key] = value;
    });

  switch (command) {
    case "airplayEnabled":
      return itunes.airplayEnabled();
    case "converting":
      return itunes.converting();
    case "currentAirPlayDevices":
      return JSON.stringify(currentAirPlayDevices(options));
    case "currentEncoder":
      return JSON.stringify(currentEncoder(options));
    case "currentEQPreset":
      return JSON.stringify(currentEQPreset(options));
    case "currentPlaylist":
      return JSON.stringify(currentPlaylist(options));
    case "currentStreamTitle":
      return itunes.currentStreamTitle();
    case "currentStreamURL":
      return itunes.converting();
    case "currentTrack":
      return JSON.stringify(currentTrack(options));
    case "currentVisual":
      return JSON.stringify(currentVisual(options));
    case "eqEnabled":
      return itunes.eqEnabled();
    case "fixedIndexing":
      return itunes.fixedIndexing();
    case "frontmost":
      return itunes.frontmost();
    case "fullScreen":
      return itunes.fullScreen();
    case "name":
      return itunes.name();
    case "mute":
      return itunes.mute();
    case "playerPosition":
      return itunes.playerPosition();
    case "playerState":
      return itunes.playerState();
    case "selection":
      return itunes.selection();
    case "shuffleEnabled":
      return itunes.shuffleEnabled();
    case "shuffleMode":
      return itunes.shuffleMode();
    case "songRepeat":
      return itunes.songRepeat();
    case "soundVolume":
      return itunes.soundVolume();
    case "version":
      return itunes.version();
    case "visualsEnabled":
      return itunes.visualsEnabled();
    case "visualSize":
      return itunes.visualSize();
    //
    case "browserWindows":
      return JSON.stringify(browserWindow(options));
    case "encoders":
      return JSON.stringify(encoders(options));
    case "eqPresets":
      return JSON.stringify(eqPresets(options));
    case "eqWindows":
      return JSON.stringify(eqWindows(options));
    case "miniplayerWindows":
      return JSON.stringify(miniplayerWindows(options));
    case "playlists":
      return JSON.stringify(playlists(options));
    case "playlistWindows":
      return JSON.stringify(playlistWindows(options));
    case "sources":
      return JSON.stringify(sources(options));
    case "tracks":
      return JSON.stringify(tracks(options));
    case "videoWindows":
      return JSON.stringify(videoWindows(options));
    case "visuals":
      return JSON.stringify(visuals(options));
    case "windows":
      return JSON.stringify(windows(options));
    case "userPlaylists":
      return JSON.stringify(userPlaylists(options));
    case "subscriptionPlaylists":
      return JSON.stringify(subscriptionPlaylists(options));
    case "libraryPlaylists":
      return JSON.stringify(libraryPlaylists(options));
    case "folderPlaylists":
      return JSON.stringify(folderPlaylists(options));
    case "audioCDPlaylists":
      return JSON.stringify(folderPlaylists(options));
    case "radioTunerPlaylist":
      return JSON.stringify(radioTunerPlaylists(options));
    case "backTrack":
      return itunes.backTrack();
    case "fastForward":
      return itunes.fastForward();
    case "nextTrack":
      return itunes.nextTrack();
    case "pause":
      return itunes.pause();
    case "playpause":
      return itunes.playpause();
    case "previousTrack":
      return itunes.previousTrack();
    case "resume":
      return itunes.resume();
    case "rewind":
      return itunes.rewind();
    case "stop":
      return itunes.stop();
    default:
      console.log("Unknown Arguments.");
      returnError();
  }
}

File: itunes/itunes.go
package itunes

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
)

func scriptExec(arg ...string) ([]byte, error) {
	cmd := switchExec(arg...)
	return cmd.Output()
}

func scriptAsyncExec(arg ...string) error {
	cmd := switchExec(arg...)
	return cmd.Start()
}

func scriptExecBool(arg ...string) (bool, error) {
	res, _ := scriptExec(arg...)
	return strconv.ParseBool(string(res))
}

func unmarshal(command string, v interface{}, arg ...string) error {
	args := []string{
		command,
	}
	args = append(args, arg...)
	jsonb, err := scriptExec(args...)
	if err != nil {
		return err
	}

	err = json.Unmarshal(jsonb, &v)
	if err != nil {
		return err
	}
	return nil
}

func trim(bString []byte) string {
	return strings.TrimSpace(string(bString))
}

// AirplayEnabled is AirPlay currently enabled?
func AirplayEnabled() bool {
	flag, _ := scriptExecBool("airplayEnabled")
	return flag
}

// Converting is a track currently being converted?
func Converting() bool {
	flag, _ := scriptExecBool("converting")
	return flag
}

// CurrentAirPlayDevices the currently selected AirPlay device(s)
func CurrentAirPlayDevices() ([]AirPlayDevice, error) {
	airPlayDevices := []AirPlayDevice{}
	err := unmarshal("currentAirPlayDevices", &airPlayDevices)
	return airPlayDevices, err
}

// CurrentEncoder the currently selected encoder (MP3, AIFF, WAV, etc.)
func CurrentEncoder() (Encoder, error) {
	encoder := Encoder{}
	err := unmarshal("currentEncoder", &encoder)
	return encoder, err
}

// CurrentEQPreset the currently selected equalizer preset
func CurrentEQPreset() (EQPreset, error) {
	eq := EQPreset{}
	err := unmarshal("currentEQPreset", &eq)
	return eq, err
}

// CurrentPlaylist the playlist containing the currently targeted track
func CurrentPlaylist() (Playlist, error) {
	playlist := Playlist{}
	err := unmarshal("currentPlaylist", &playlist)
	return playlist, err
}

// CurrentStreamTitle the name of the current song in the playing stream (provided by streaming server)
func CurrentStreamTitle() string {
	res, _ := scriptExec("currentStreamTitle")
	return trim(res)
}

// CurrentStreamURL the URL of the playing stream or streaming web site (provided by streaming server)
func CurrentStreamURL() string {
	res, _ := scriptExec("currentStreamURL")
	return trim(res)
}

// CurrentTrack the current targeted track
func CurrentTrack(artwork bool) (Track, error) {
	track := Track{}
	args := []string{
		fmt.Sprintf("artwork=%t", artwork),
	}

	err := unmarshal("currentTrack", &track, args...)
	return track, err
}

// CurrentVisual the currently selected visual plug-in
func CurrentVisual() (Visual, error) {
	visual := Visual{}
	err := unmarshal("currentVisual", &visual)
	return visual, err
}

// EqEnabled is the equalizer enabled?
func EqEnabled() bool {
	flag, _ := scriptExecBool("eqEnabled")
	return flag
}

// FixedIndexing true if all AppleScript track indices should be independent of the play order of the owning playlist.
func FixedIndexing() bool {
	flag, _ := scriptExecBool("fixedIndexing")
	return flag
}

// Frontmost is iTunes the frontmost application?
func Frontmost() bool {
	flag, _ := scriptExecBool("frontmost")
	return flag
}

// FullScreen are visuals displayed using the entire screen?
func FullScreen() bool {
	flag, _ := scriptExecBool("fullScreen")
	return flag
}

// Name the name of the application
func Name() string {
	res, _ := scriptExec("name")
	return trim(res)
}

// Mute has the sound output been muted?
func Mute() bool {
	flag, _ := scriptExecBool("mute")
	return flag
}

// PlayerPosition the player’s position within the currently playing track in seconds.
func PlayerPosition() float64 {
	res, _ := scriptExec("playerPosition")
	position, _ := strconv.ParseFloat(strings.TrimSpace(string(res)), 64)
	return position
}

// PlayerState is iTunes stopped, paused, or playing?
func PlayerState() string {
	res, _ := scriptExec("playerState")
	return trim(res)
}

// Selection the selection visible to the user
func Selection() string {
	res, _ := scriptExec("selection")
	return trim(res)
}

// ShuffleEnabled are songs played in random order?
func ShuffleEnabled() bool {
	flag, _ := scriptExecBool("shuffleEnabled")
	return flag
}

// ShuffleMode the playback shuffle mode
func ShuffleMode() string {
	res, _ := scriptExec("shuffleMode")
	return trim(res)
}

// SongRepeat the playback repeat mode
func SongRepeat() string {
	res, _ := scriptExec("songRepeat")
	return trim(res)
}

// SoundVolume the sound output volume (0 = minimum, 100 = maximum)
func SoundVolume() int {
	res, _ := scriptExec("soundVolume")
	volume, _ := strconv.Atoi(string(res))
	return volume
}

// Version the version of iTunes
func Version() string {
	res, _ := scriptExec("version")
	return trim(res)
}

// VisualsEnabled are visuals currently being displayed?
func VisualsEnabled() bool {
	flag, _ := scriptExecBool("visualsEnabled")
	return flag
}

// VisualSize the size of the displayed visual
func VisualSize() string {
	res, _ := scriptExec("visualSize")
	return trim(res)
}

// AirplayDevices AirPlay devices
func AirplayDevices() ([]AirPlayDevice, error) {
	airPlayDevices := []AirPlayDevice{}
	err := unmarshal("airplayDevices", &airPlayDevices)
	return airPlayDevices, err
}

// BrowserWindows main iTunes windows
func BrowserWindows() ([]BrowserWindow, error) {
	browserWindows := []BrowserWindow{}
	err := unmarshal("browserWindows", &browserWindows)
	return browserWindows, err
}

// Encoders converts a track to a specific file formats
func Encoders() ([]Encoder, error) {
	encoders := []Encoder{}
	err := unmarshal("encoders", &encoders)
	return encoders, err
}

// EqPresets equalizer preset configurations
func EqPresets() ([]EQPreset, error) {
	eqPresets := []EQPreset{}
	err := unmarshal("eqPresets", &eqPresets)
	return eqPresets, err
}

// EqWindows iTunes equalizer windows
func EqWindows() ([]EQWindow, error) {
	eqWindows := []EQWindow{}
	err := unmarshal("eqPresets", &eqWindows)
	return eqWindows, err
}

// MiniplayerWindows miniplayer windows
func MiniplayerWindows() ([]MiniplayerWindow, error) {
	miniplayerWindows := []MiniplayerWindow{}
	err := unmarshal("miniplayerWindows", &miniplayerWindows)
	return miniplayerWindows, err
}

// Playlists list of songs/streams
func Playlists() ([]Playlist, error) {
	playlists := []Playlist{}
	err := unmarshal("playlists", &playlists)
	return playlists, err
}

// PlaylistWindows sub-window showing a single playlist windows
func PlaylistWindows() ([]PlaylistWindow, error) {
	playlistWindows := []PlaylistWindow{}
	err := unmarshal("playlistWindows", &playlistWindows)
	return playlistWindows, err
}

// Sources music sources (music library, CD, device, etc.)
func Sources() ([]Source, error) {
	sources := []Source{}
	err := unmarshal("sources", &sources)
	return sources, err
}

// Tracks playable audio sources
func Tracks() ([]Track, error) {
	tracks := []Track{}
	err := unmarshal("tracks", &tracks)
	return tracks, err
}

// VideoWindows VideoWindows
func VideoWindows() ([]VideoWindow, error) {
	videoWindows := []VideoWindow{}
	err := unmarshal("videoWindows", &videoWindows)
	return videoWindows, err
}

// Visuals visual plug-in list
func Visuals() ([]Visual, error) {
	visuals := []Visual{}
	err := unmarshal("visuals", &visuals)
	return visuals, err
}

// Windows the currently windows
func Windows() ([]Window, error) {
	windows := []Window{}
	err := unmarshal("windows", &windows)
	return windows, err
}

// // UserPlaylists custom playlists created by the user
// func UserPlaylists() ([]UserPlaylist, error) {
// 	userPlaylists := []UserPlaylist{}
// 	err := unmarshal("userPlaylists", &userPlaylists)
// 	return userPlaylists, err
// }

// // SubscriptionPlaylist subscription playlists from Apple Music
// func SubscriptionPlaylists() ([]SubscriptionPlaylist, error) {
// 	subscriptionPlaylists := []SubscriptionPlaylist{}
// 	err := unmarshal("subscriptionPlaylists", &subscriptionPlaylists)
// 	return subscriptionPlaylists, err
// }

// // LibraryPlaylists the master music library playlists
// func LibraryPlaylists() ([]LibraryPlaylist, error) {
// 	libraryPlaylists := []LibraryPlaylist{}
// 	err := unmarshal("libraryPlaylists", &libraryPlaylists)
// 	return libraryPlaylists, err
// }

// // FolderPlaylists folder that contains other playlists
// func FolderPlaylists() ([]FolderPlaylist, error) {
// 	folderPlaylist := []FolderPlaylist{}
// 	err := unmarshal("folderPlaylists", &folderPlaylist)
// 	return folderPlaylist, err
// }

// // AudioCDPlaylists playlist representing an audio CD
// func AudioCDPlaylists() ([]AudioCDPlaylist, error) {
// 	audioCDPlaylist := []AudioCDPlaylist{}
// 	err := unmarshal("audioCDPlaylists", &audioCDPlaylist)
// 	return audioCDPlaylist, err
// }

// // RadioTunerPlaylist the radio tuner playlist
// func RadioTunerPlaylists() ([]RadioTunerPlaylist, error) {
// 	radioTunerPlaylist := []RadioTunerPlaylist{}
// 	err := unmarshal("radioTunerPlaylist", &radioTunerPlaylist)
// 	return radioTunerPlaylist, err
// }

// Controll

// BackTrack reposition to beginning of current track or go to previous track if already at start of current track
func BackTrack() {
	scriptAsyncExec("backTrack")
}

// FastForward skip forward in a playing track
func FastForward() {
	scriptAsyncExec("fastForward")
}

// NextTrack advance to the next track in the current playlist
func NextTrack() {
	scriptAsyncExec("nextTrack")
}

// Pause pause playback
func Pause() {
	scriptAsyncExec("pause")
}

// Playpause toggle the playing/paused state of the current track
func Playpause() {
	scriptAsyncExec("playpause")
}

// PreviousTrack return to the previous track in the current playlist
func PreviousTrack() {
	scriptAsyncExec("previousTrack")
}

// Resume disable fast forward/rewind and resume playback, if playing.
func Resume() {
	scriptAsyncExec("resume")
}

// Rewind skip backwards in a playing track
func Rewind() {
	scriptAsyncExec("rewind")
}

// Stop stop playback
func Stop() {
	scriptAsyncExec("stop")
}
